modules/protect/blocked-login-page.php

```diff
@@ -0,0 +1,337 @@
+		if ( $this->valid_blocked_user_id && $this->valid_blocked_user_id != $user->ID ) {
+			return new WP_Error( 'invalid_recovery_token', __( 'The recovery token is not valid for this user.', 'jetpack' ) );
+		}
+
+		return $user;
+	}
+
+	public function is_blocked_user_valid() {
+		if ( ! $this->can_send_recovery_emails ) {
+			return false;
+		}
+
+		if ( $this->valid_blocked_user_id ) {
+			return true;
+		}
+
+		if ( ! isset( $_GET['validate_jetpack_protect_recovery'], $_GET['user_id'] ) ) {
+			return false;
+		}
+
+		if ( ! $this->is_valid_protect_recovery_key( $_GET['validate_jetpack_protect_recovery'], $_GET['user_id'] ) ) {
+			return false;
+		}
+
+		$this->valid_blocked_user_id = (int) $_GET['user_id'];
+
+		return true;
+	}
+
+	public function is_valid_protect_recovery_key( $key, $user_id ) {
+
+		$path     = sprintf( '/sites/%d/protect/recovery/confirm', Jetpack::get_option( 'id' ) );
+		$response = Jetpack_Client::wpcom_json_api_request_as_blog(
+			$path,
+			'1.1',
+			array(
+				'method' => 'post'
+			),
+			array(
+				'token'   => $key,
+				'user_id' => $user_id,
+				'ip'      => $this->ip_address,
+			)
+		);
+
+		$result = json_decode( wp_remote_retrieve_body( $response ) );
+
+		if ( is_wp_error( $result ) || empty( $result ) || isset( $result->error ) ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	public function render_and_die() {
+		$this->page_title = __( 'Login Blocked by Jetpack', 'jetpack' );
+
+		if ( ! $this->can_send_recovery_emails ) {
+			$this->render_blocked_login_message();
+
+			return;
+		}
+
+		if ( isset( $_GET['validate_jetpack_protect_recovery'] ) && $_GET['user_id'] ) {
+			$this->protect_die( __( 'Could not validate recovery token.', 'jetpack' ) );
+
+			return;
+		}
+
+		if (
+			isset( $_GET['jetpack-protect-recovery'] ) &&
+			isset( $_POST['_wpnonce'] ) &&
+			wp_verify_nonce( $_POST['_wpnonce'], 'bypass-protect' )
+		) {
+			$this->process_recovery_email();
+
+			return;
+		}
+
+		if ( isset( $_GET['loggedout'] ) && 'true' === $_GET['loggedout'] ) {
+			$this->protect_die( __( 'You successfully logged out.', 'jetpack' ) );
+		}
+
+		$this->render_recovery_form();
+	}
+
+	public function render_blocked_login_message() {
+		$this->protect_die( $this->get_html_blocked_login_message() );
+	}
+
+	function process_recovery_email() {
+		$sent = $this->send_recovery_email();
+
+		if ( is_wp_error( $sent ) ) {
+			$this->protect_die( $sent, true );
+		} else {
+			$this->render_recovery_success();
+		}
+	}
+
+	function send_recovery_email() {
+		$email = isset( $_POST['email'] ) ? $_POST['email'] : '';
+		if ( sanitize_email( $email ) !== $email || ! is_email( $email ) ) {
+			return new WP_Error( 'invalid_email', __( "Oops, looks like that's not the right email address. Please try again!", 'jetpack' ) );
+		}
+		$user = get_user_by( 'email', trim( $email ) );
+
+		if ( ! $user ) {
+			return new WP_Error( 'invalid_user', __( 'Oops, could not find a user with that email address.', 'jetpack' ) );
```

Comment:
Could this somehow be used by a hacker who is already ip blocked to test for valid email addresses/accounts on the site/domain?